clients/src/main/java/org/apache/kafka/common/record/ByteBufferOutputStream.java

```diff
@@ -16,54 +16,82 @@
  */
 package org.apache.kafka.common.record;
 
-import java.io.DataOutputStream;
 import java.io.OutputStream;
 import java.nio.ByteBuffer;
 
 /**
- * A byte buffer backed output outputStream
+ * A ByteBuffer-backed OutputStream
  */
-public class ByteBufferOutputStream extends DataOutputStream {
+public class ByteBufferOutputStream extends OutputStream {
 
     private static final float REALLOCATION_FACTOR = 1.1f;
 
+    private ByteBuffer buffer;
+    private int initialPosition;
+
     public ByteBufferOutputStream(ByteBuffer buffer) {
-        super(new UnderlyingOutputStream(buffer));
+        this.buffer = buffer;
+        this.initialPosition = buffer.position();
     }
 
-    public ByteBuffer buffer() {
-        return ((UnderlyingOutputStream) out).buffer;
+    public void write(int b) {
+        if (buffer.remaining() < 1)
+            expandBuffer(buffer.capacity() + 1);
+        buffer.put((byte) b);
     }
 
-    public static class UnderlyingOutputStream extends OutputStream {
-        private ByteBuffer buffer;
+    public void write(byte[] bytes, int off, int len) {
+        if (buffer.remaining() < len)
+            expandBuffer(buffer.position() + len);
+        buffer.put(bytes, off, len);
+    }
 
-        public UnderlyingOutputStream(ByteBuffer buffer) {
-            this.buffer = buffer;
+    public void write(ByteBuffer buffer) {
+        if (buffer.hasArray())
+            write(buffer.array(), buffer.arrayOffset() + buffer.position(), buffer.remaining());
+        else {
+            int pos = buffer.position();
+            for (int i = pos, limit = buffer.remaining() + pos; i < limit; i++)
+                write(buffer.get(i));
```

Comment:
Similar to the other comment, we should just use the bulk `put` method. That also updates the position of the received buffer, which is expected in cases like this.